### PR TITLE
python2Packages.certifi: init at 2019.11.28

### LINKED ISSUE #127423: nixops-1.7: no longer builds on master and 21.05
**Describe the bug**
The nixops-1.7 package is failing to build on NixOS 21.05 after a recent upgrade to the python certifi package (PR #125996) enabled a test that fails with python2. The certifi package is an indirect dependency of nixops-1.7.

**To Reproduce**
The failing build can be viewed on hydra: https://hydra.nixos.org/build/145417618

**Expected behavior**
Ideally, the nixops-1.7 package will continue to build and remain available until a newer release of nixops is available.

**Notify maintainers**

@AmineChikhaoui @edolstra @rbvermaa

CC: @fabaff @dotlambda @adisbladis 

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.37, NixOS, 21.05.git.729e236f7a0 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.12`
 - channels(root): `""`
 - nixpkgs: `/etc/nixos/nixpkgs`
 
NOTE: nixops-1.7 is still available when running with the system configuration above. The failing build is visible on hydra, as described in the "To Reproduce" section above.

I plan to file a PR that will disable the python certifi tests for python2 in pkgs/development/python-modules/certifi/default.nix. The PR will probably need to be filed against staging and require a backport to 21.05, if accepted.

As the certifi tests were just enabled, removing them for python2 is unlikely to introduce new issues. The certifi library provides two functions: where() and contents(). The contents() function fails in python2. However, the where() function still seems to work and appears to be the function used by the python requests library (used by boto).

As a note, the underlying issue with python2 compatibility seems to predate the certifi version bump in PR #125996. The newer certifi is primarily an update to the certificate file:

```
$ git diff --numstat 2020.12.05 2021.05.30
1       1       LICENSE
1       1       certifi/__init__.py
230     298     certifi/cacert.pem
```
See comparison here: https://github.com/certifi/python-certifi/compare/2020.12.05...2021.05.30

Thanks to @dotlambda for helping me to better understand this issue on the NixOS Dev matrix yesterday.

Maintainer information:
```yaml
attribute: nixops
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by dotlambda**
###### Motivation for this change
fixes https://github.com/NixOS/nixpkgs/issues/127423

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @tollb @NixOS/nixops-committers

### CODE REVIEW

**Comment by dotlambda** [comment]
To be honest, I don't like providing an outdated version of certifi for Python 2 because we shouldn't ship outdated security-related stuff. Maybe it's better not to fix this and mark NixOps 1.7 as broken.
Also, I will unsubscribe from this pull request. Any necessary modifications can be force-pushed to my branch by @NixOS/nixops-committers who should also merge this pull request if they think it's the best solution or create a new one otherwise.
This has been the very last time I put any work into fixing some Python 2 package required by NixOps!

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 127453` at 304839a9 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>11 packages marked as broken and skipped:</summary>
  <ul>
    <li>gnuradio3_7</li>
    <li>gnuradio3_7Packages.ais</li>
    <li>gnuradio3_7Packages.gnuradio</li>
    <li>gnuradio3_7Packages.gsm</li>
    <li>gnuradio3_7Packages.limesdr</li>
    <li>gnuradio3_7Packages.nacl</li>
    <li>gnuradio3_7Packages.osmosdr</li>
    <li>gnuradio3_7Packages.rds</li>
    <li>natron</li>
    <li>ocropus</li>
    <li>rainbowstream</li>
  </ul>
</details>
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/24f8fbde2f47904deb7e2ea9ac7921f3">graalvm8</a></li>
    <li><a href="https://gist.github.com/e1165e6f134c438729992f3dda44e45f">jvmci8</a></li>
    <li><a href="https://gist.github.com/5baf0e8e9446163263e6be90ba610e62">opae</a></li>
  </ul>
</details>
<details>
  <summary>12 packages built successfully:</summary>
  <ul>
    <li>bitbucket-cli</li>
    <li>buttersink</li>
    <li>cloudmonkey</li>
    <li>datadog-agent</li>
    <li>euca2ools</li>
    <li>fmbt</li>
    <li>git-fast-export</li>
    <li>mercurial_4</li>
    <li>mx</li>
    <li>nixops</li>
    <li>tsung</li>
    <li>zabbix-cli</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

**Comment by mweinelt** [comment]
This poses a certain security risk, so I decided to set `knownVulnerabilities`.

File: pkgs/development/python-modules/certifi/2019.11.28.nix
Comment (Ekleog) [inline pkgs/development/python-modules/certifi/2019.11.28.nix:26]: Can anyone describe an actual threat model based on “shipping an outdated certificate store”? I can't think of any that would actually be a problem for normal users, and non-normal users already have much worse to worry about.

IMO adding `knownVulnerabilities` when there are no actually known vulnerabilities only makes people more used to working around it, especially for a package as advertised as nixops, and so is a net negative for the security of NixOS users.

Now, if there were some specific known vulnerability (eg. “this CA that was there in 2019 was removed after having issued malicious certificates that can still be used”), I would totally understand this item; but IMO as currently described this is lacking motivation and it'd be better to have no knownVulnerabilities than this one.

**Review by Ekleog — CHANGES_REQUESTED** [review]

**Comment by nlewo** [comment]
@dotlambda Note this would also fixes the `python2Packages.requests` library which depends on `certifi`.

File: pkgs/development/python-modules/certifi/2019.11.28.nix
Comment (zeri42) [reply pkgs/development/python-modules/certifi/2019.11.28.nix:26]: easy. Some CA get's their private key stolen and now the thief can issue new certificates. The answer to this is revocation. unless you have OCSP revocation is done via two ways:
1. certificate revocation lists
2. removing the CA from certificate stores like this one
Hence keeping the certificate store up to date is important. For reference: https://sslmate.com/certspotter/failures

Comment (aszlig) [reply pkgs/development/python-modules/certifi/2019.11.28.nix:26]: Hm, what about re-using the bundle from the `cacert` package instead? This would not only solve this issue but also would consolidate the CA bundle to a single location. For example when writing integration tests involving Python software, one usually needs to not only override `cacert` but also `python*Packages.certifi` to inject testing certificates.

Comment (tollb) [reply pkgs/development/python-modules/certifi/2019.11.28.nix:26]: The approach suggested by @aszlig may also be applicable to the boto 2.49.0 package, as I believe it has an embedded cacert file that is even further out-of-date: https://raw.githubusercontent.com/boto/boto/2.49.0/boto/cacerts/cacerts.txt. I'm not certain when this embedded cacerts.txt is used by boto versus the one provided by python2Packages.certifi.

If someone can confirm that boto's embedded cacerts.txt is actually used (especially through nixops 1.7), we might want to add a knownVulnerabilities to the boto package as well -- for consistency with python2Packages.certifi. Or, provide updated certificates in both packages. I took a quick look at arch and debian to see if they had updated the certs in their boto 2.49.0 packages, and it doesn't seem that they have. So, maybe I am mistaken.

**Comment by dpausp** [comment]
We could also use a recent version of certifi and disable the test for the `contents()` function. The function is broken but it didn't even exist on 2019.11.28. This wouldn't break existing Python 2.7 code where the new function has never worked. Ugly, but having newer certificates is better, I think.

File: pkgs/development/python-modules/certifi/2019.11.28.nix
Comment (Ekleog) [reply pkgs/development/python-modules/certifi/2019.11.28.nix:26]: @zeri42 Thank you, your link confirms that as of today there is no known instance of a CA having an issue that would warrant the `knownVulnerabilities` toggle since the version of certifi that is being considered here, and so that reverting updates should not be a security vulnerability :) (though it is only one source, of course)

(And also, even if they were in most cases I'd keep my position: an attacker stealing the private key from a CA is not a reasonable threat model for a normal NixOS user, which is the reason why I wrote my first message)

This being said I'm going to step out of this discussion: I'm only focussed on “`knownVulnerabilities` here would be bad”; there are for sure much better things that could be done by improving stuff in various ways :)

**Comment by tomfitzhenry** [comment]
This will also fix buttersink failing to build: https://github.com/NixOS/nixpkgs/issues/128151.

**Comment by prusnak** [comment]
This will also fix gnuradio 3.7 shipped with nixos-21.05: https://github.com/NixOS/nixpkgs/issues/128620

**Comment by adisbladis** [comment]
@dotlambda I took the liberty of reworking this PR a bit.
We now use `2019.11.28` for building the package but inherit the cert bundle from the maintained python3-only version.

I think this is a decent solution with minimal maintenance.

**Comment by prusnak** [comment]
@adisbladis I wonder whether the package version of this hybrid should be `2019.11.28` (source code version as reported by `__init__.py`) or `2021.05.30` (the version of `cacert.pem`)

**Comment by dotlambda** [comment]
The version should be 2019.11.28 because the API has since been changed.

File: pkgs/development/python-modules/certifi/python2.nix
Comment (prusnak) [inline pkgs/development/python-modules/certifi/python2.nix:16]: Please use the original version:

```suggestion
  version = "2019.11.28";

  src = fetchPypi {
    inherit pname version;
```

**Review by prusnak — CHANGES_REQUESTED** [review]

**Comment by adisbladis** [comment]
Alright, version set as `2019.11.28`.

**Comment by clefru** [comment]
https://github.com/clefru/nixpkgs/commit/57e65ad7e61dfa690bfa7eb35c13f9787d70f32e has a commit that patches Python 2 support back into a recent certifi. Works for me so far.

**Comment by adisbladis** [comment]
@clefru I thought about adopting a similar approach but opted against it because of the added maintenance burden.

I think for now I'll just merge this and if you think your approach is sufficiently better could you please create a separate PR?

**Pull request merged**
